### ISSUE TITLE
ManyToOne with EntityType by_reference=false results in "Entities passed to the choice field must be managed. Maybe persist them in the entity manager?"

### ISSUE THREAD

**Issue by smcjones** [symfony/symfony#23338]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.3.0

This is part bug report for what may be an extreme edge case, and part notice for anyone else who might be struggling with debugging this issue.

I have a custom script which loads rules from a form, and there was an error with it–not an issue with Symfony per se–whereby I was passing everything with `by_reference=false`. This had unintended consequences and resulted in a cryptic exception message from Symfony:

    Entities passed to the choice field must be managed. Maybe persist them in the entity manager?

I have been able to determine that this only happens when `by_reference` is passed as false.

Like in #13964, if I remove the offending Exception everything works as anticipated. 

In hindsight, it makes a lot of sense that `by_reference=false` would cause this error because, as I understand it, this is considered an entirely separate entity. For this reason, I'm not sure if there is anything to "fix", but hopefully if it's not something to be fixed, this can be more easily specified, saving people hours of debugging time, either in an exception message or documentation as a known issue.

At the very least, if someone is struggling with this somewhat challenging bug, it's another avenue to test against, and perhaps the documentation can be updated as a known issue for passing by_reference? 

Thanks,

Sean

**Comment by HeahDude** [comment]
Hello @smcjones and thank you for reporting this.

What would be the reason to use `by_reference` as `false` here?

As I understand it a ManyToOne relation implies your underlying data has "one" nested entity mapped to the `EntityType`. Whereas such option should be used with a collection of objects (ManyToMany or OneToMany, using then the `multiple` option as `true`) and just in the case you want to force "setter/adder/remover" to be called because a Doctrine `Collection` is an object so is modified by reference (which prevent the Form from calling those methods).

Looks like a doc issue to me.

Am I missing something?

**Comment by smcjones** [comment]
There is actually no reason to set by_reference to false. I am writing a fairly large bundle that seeks to bridge the gap between Symfony and HTML_QuickForm for legacy applications. At first I set "by_reference" to false. I think this is an extreme edge case, but it is still worth noting. I agree that it is probably a doc issue (as I think I noted in my original post).

That being said, it does seem that the exception pointed out in #13964 is a little too broad. It appears that there are times when simply commenting out the Exception makes everything work perfectly. Perhaps having some additional checks before triggering an Exception (such as when an entity is passed `by_reference`) could be helpful. At the very least, it would be nice to have a more expressive Exception message (passing entities `by_reference` is not supported for `persist` operations) if in fact this is intentional.

**Comment by smcjones** [comment]
> I'm not sure I'm following. by_reference option is true by default, so it implies the dev explicitly set it to false and that implies he "knows" why he's doing so.

There are reasons to set `by_reference` to false, as marked in the documentation. However, and I will try to stress this, **there is no reference at all to the behavior I have outlined.** I am not saying that the behavior I outlined is a bug. What I am saying is that, if it's expected, it should be documented.

> This is not true, it is fine to have nested entities during persist operations as long as a cascade is set in the mapping.

I hoped it would be obvious that I tried to set cascade="persist" by the fact that commenting out the Exception allowed my data to persist. I am assuming that perhaps you have not tried to replicate this behavior based on your response.
